en/lessons/advanced/otp-distribution.md

```diff
@@ -0,0 +1,430 @@
+
+## Distributed Tasks
+
+[Distributed tasks](https://hexdocs.pm/elixir/master/Task.html#module-distributed-tasks) allow us to spawn supervised tasks across nodes. We'll build a simple supervisor application that leverages distributed tasks to allow users to chat with one another via an `iex` session, across distributed nodes.
+
+### Defining the Supervisor Application
+Generate your app:
+
+```
+mix new chat --sup
+```
+
+### Adding the Task Supervisor to the Supervision Tree
+A Task Supervisor dynamically supervises tasks. It is started with no children, often _under_ a supervisor of its own, and can can be used later on to supervise any number of tasks.
+
+We'll add a Task Supervisor to our app's supervision tree and name it `Chat.TaskSupervisor`
+
+```elixir
+# lib/chat/application.ex
+defmodule Chat.Application do
+  @moduledoc false
+
+  use Application
+
+  def start(_type, _args) do
+    children = [
+      {Task.Supervisor, name: Chat.TaskSupervisor}
+    ]
+
+    opts = [strategy: :one_for_one, name: Chat.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+
+Now we know that wherever our application is started on a given node, the `Chat.Supervisor` is running and ready to supervise tasks.
+
+### Sending Messages with Supervised Tasks
+
+We'll start supervised tasks with the [`Task.Supervisor.async/4`](https://hexdocs.pm/elixir/master/Task.Supervisor.html#async/5) function.
+
+This function takes in four arguments:
+
+* The supervisor we want to use to supervise the task. This can be passed in as a tuple of `{SupervisorName, remote_node_name}` in order to supervise the task on the remote node.
+* The name of the module on which we want to execute a function
+* The name of the function we want to execute
+* Any arguments that need to be supplied to that function
+
+Out Chat application is pretty simple. It sends messages to remote nodes and remote nodes respond to those messages by `IO.puts`-ing them out to the STDOUT of the remote node.
+
+First, let's define a function, `receive_message`, that we want our task to execute on a remote node.
+
+```elixir
+# lib/chat.ex
+defmodule Chat do
+  def receive_message(message) do
+    IO.puts message
+  end
+end
+```
+
+Next up, let's teach the `Chat` module how to send the message to a remote node using a supervised task. We'll define a method `send_message` that will enact this process:
+
+```elixir
+# lib/chat.ex
+defmodule Chat do
+  ...
+
+  def send_message(recipient, message) do
+    spawn_task(__MODULE__, :receive_message, recipient, [message])
+  end
+
+  def spawn_task(module, fun, recipient, args) do
+    Task.Supervisor.async(remote_supervisor(recipient), module, fun, args)
```

Comment:
Our code examples follow the [community styleguide](https://github.com/christopheradams/elixir_style_guide) for the most part, so is covered under starting pipelines with a bare value [here](https://github.com/christopheradams/elixir_style_guide#bare-variables).

We can refactor this to something like this:

```elixir
recipient
|> remote_supervisor()
|> Task.Supervisor.async(module, fun, args)
|> Task.await
```